File: 프로그래머스/2/388352. 비밀 코드 해독/비밀 코드 해독.java
import java.util.*;
import java.util.stream.Collectors;

class Solution {
    
    static List<Set<Integer>> combinationList = new ArrayList<>();
    
    public int solution(int n, int[][] q, int[] ans) {
        int answer = 0;
        combination(new int[n], 0, n, 5, 0);
        List<Set<Integer>> qList = new ArrayList<>();
          for (int[] ints : q) {
            Set<Integer> ls = Arrays.stream(ints).boxed().collect(Collectors.toSet());
            qList.add(ls);
          }

        for(Set<Integer> comb : combinationList){
            boolean flag = true;
//            System.out.println(comb);
            for(int i = 0 ; i < q.length ; i++){
                int count = 0;
                for (Integer element : comb) {
                    if (qList.get(i).contains(element)) {
                        count++;
                    }
                }
                if(count != ans[i]){
                    flag = false;
                    break;
                }
            }
            if(flag){
                answer++;
            }

        }
        return answer;
    }
    
    public void combination(int[] arr, int index, int n, int r, int target) {
    if (r == 0){
        Set<Integer> ls = Arrays.stream(arr).boxed().collect(Collectors.toSet());
        combinationList.add(ls);
    } else if (target == n){
        return;
    } else {
        arr[index] = target+1;
        combination(arr, index + 1, n, r - 1, target + 1);
        combination(arr, index, n, r, target + 1);
    }
    }
}

File: 프로그래머스/2/388352. 비밀 코드 해독/비밀 코드 해독.kt
import java.util.*
class Solution {
fun solution(n: Int, q: Array<IntArray>, ans: IntArray): Int {
    var answer: Int = 0
    val possible = (1..n).toSet().toIntArray()
    val result = mutableSetOf<Set<Int>>()

    for(com in combination(possible,5)){

        var isPossible = true
        for(i in q.indices){
            if(q[i].toSet().intersect(com).size != ans[i]){
                isPossible = false
                break
            }
        }
        if(isPossible){
            answer += 1
        }
    }
    return answer
}

    fun  combination(elements: IntArray, r: Int): List<Set<Int>> {
        val n = elements.size
        val results = mutableListOf<Set<Int>>() // 모든 경우의 수

        val result = elements.sliceArray(0 until r)

        fun recursionCombination(depth: Int = 0, index: Int = 0) {
            if (depth == r) {
                results.add(result.toSet())
                return
            }

            for (i in index until n) {
                result[depth] = elements[i]
                recursionCombination(depth + 1, i + 1)
            }
        }

        recursionCombination()
        return results
    }

}